src/App.js

```diff
@@ -34,8 +34,15 @@ export default function App({
     return createElement(PageContext.Provider, { value: page.props }, null)
   }
 
-  const renderChildren = children
-    || (({ Component, props, key }) => createElement(Component, { key, ...props }))
+  const renderChildren = children || (({ Component, props, key }) => {
+    const child = createElement(Component, { key, ...props })
+
+    if (Component.layout) {
+      return Component.layout(child)
```

Comment:
Right now layout is just a function, I believe we should upgrade it to a component so it has a lifecycle of its own, and can accept any generic component.

```suggestion
      return createElement(Component.Layout, { children })
```
The example would then look like:

```js
Home.Layout = ({ children }) => <Layout children={children} title="Home" />
```

And in some cases could be simplified to:

```js
Home.Layout = Layout
```

Reply:
This would break persistent layouts because React would no longer be re-rendering the same component, every page would be creating a new on-the-fly component and React would trash the children and re-render them from scratch.

Has to just be a regular function, it's not meant to be a component 👍 

All this work is based on ideas I wrote up in more detail here:

https://adamwathan.me/2019/10/17/persistent-layout-patterns-in-nextjs/

Reply:
I can confirm this. Just tried it, and persistent layouts don't work using this approach.

Reply:
Played around with this and I get it, makes sense this way 👍